70/ETH-SECPM_improvements (#630)
vacp2p/rfc: content/docs/rfcs/70/README.md

---
slug: 70
title: 70/ETH-SECPM

> name: Secure channel setup using Ethereum accounts

Reviewer: Note:
We can either specify the messaging layer in this document as well, or handle the messaging in a separate document.
I'd suggest to cover both in this document for now and later split out the messaging part if desired.

The messaging part should also cover message reliability, because the specified protocol should work over unreliable P2P networks like Waku relay.

---

# Abstract
This specification outlines a private messaging service utilizing the Ethereum blockchain. 
Rooted in the existing [model](https://rfc.vac.dev/spec/20/), this proposal addresses the deficiencies related to forward privacy and authentication inherent in the current framework.
The specification is divided into 3 sections:

> - Private 1-to-1 communications protocol, based in Signal's double ratchet.

Reviewer: link to double ratchet

> - Private 1-to-1 communications protocol, based in Signal's double ratchet.

Reviewer: ```suggestion
- Private 1-to-1 communications protocol, based on Signal's double ratchet.
```

> - Private group messaging protocol, based in the MLS protocol.

Reviewer: ```suggestion
- Private group messaging protocol, based on the MLS protocol.
```

> - Description of an Ethereun-based authentication protocol, based in SIWE.

Reviewer: ```suggestion
- Description of an Ethereum-based authentication protocol, based on SIWE.
```

> Please note that this document is a work in progress, with forthcoming sections detailing the implementation of various functions and algorithms within the framework of the Noise protocol.

Reviewer: ```suggestion
The following section on the private 1-to-1 communications protocol specifies various functions and algorithms within the framework of the Noise protocol.
```

We should update

Reviewer: The work in progress character is already expressed by the RFC being in raw state.
If parts are especially WiP/experimental, we can mark them accordingly.

3.  Execution of the double ratchet algorithm for forward secure, authenticated communications, using the common secret key SK, obtained from X3DH, as a root key.

## Cryptographic functions required
-   XEd448 and SHA512 for digital signatures involved in the X3DH key generation.
-   SHA256 for the key generation related to AES256-GCM.
-   AES256-GCM for the encryption/decryption of messages.

### Considerations on the X3DH initialization
This scheme MUST work on a specific elliptic curves which differ from those used by Ethereum. The curve Curve448 MUST be chosen: 
since it offers a higher security level: 224-bit security instead of the 128-bit security provided by X25519. This document specifies X3DH in terms of the Noise Framework.
The X3DH algorithm corresponds to the IX pattern in Noise.

Bob and Alice MUST define personal key pairs `(ik_B, IK_B)` and `(ik_A, IK_A)` respectively where:
-   The key ik must be kept secret,
-   and the key IK is public.

Bob will not be able to use his Ethereum public key during this stage due to incompatibilities with the involved elliptic curves, therefore he MUST generate new keys. 

> Using the Noise framework notation, these key pairs MUST BE generated using `(ik_B, IK_B) = GENERATE_KEYPAIR(curve = curve448)`.

Reviewer: ```suggestion
Using the Noise framework notation, these key pairs MUST be generated using `(ik_B, IK_B) = GENERATE_KEYPAIR(curve = curve448)`.
```

In any case, the user stores the associated IPFS hash, URL or reference in Ethereum.

The fact of a user not updating the ephemeral information can be understood as Bob not willing to participate in any communication.

This applies to `KeyPackage`, which in the MLS specification are meant to be stored in a directory provided by the delivery service.
If such an element does not exist, `KeyPackage` MUST be stored according to one of the two options outlined above.

### Interaction with Ethereum

Storing static data is done using a dedicated smart contract `PublicKeyStorage` which associates the Ethereum wallet address of a user with his public key.
This mapping is done by `PublicKeyStorage` using a `publicKeys` function, or a `setPublicKey` function.
This mapping is done if the user passed an authorization process.
A user who wants to retrieve a public key associated with a specific wallet address calls a function `getPublicKey`.
The user provides the wallet address as the only input parameter for `getPublicKey`.
The function outputs the associated public key from the smart contract.

> # Extension to group chat: the MLS Framework

Reviewer: ```suggestion
# Extension to Group Chat
```

> 

Reviewer: one intro sentence making explicit that the following in based on the RFC9420; otherwise it might read we are proposing MLS as something novel.

> The Messaging Layer Security (MLS) protocol aims at providing a group of users with end-to-end encryption in an authenticated and asynchronous way. 

Reviewer: We should link the RFC here.

# Extension to group chat

## 1-to-1 version: introduction

> In order to extend the protocol to a group chat, this document specifies using an Asynchronous Distributed Key Generation (ADKG) to replace the X3DH step in the previous combination X3DH + double ratchet.

Reviewer: > using an Asynchronous Distributed Key Generation (ADKG)

Is this still up-to-date?

whereas high-threshold protocols admit reconstruction thresholds much higher than the number of malicious nodes.

Existing ADKG constructions tend to become inefficient when the reconstruction threshold surpasses one-third of the total nodes. 

> In this proposal we suggest using the scheme by [Kokoris-Kogias et al.](https://eprint.iacr.org/2022/1389) which is designed for `n = 3t + 1` nodes. 

Reviewer: In an RFC, we should work towards clearly specifcying which scheme should/must be used.
Not sure if the ADKG part is relevant still.
